File: sdks/go/credential.go
package boxlite

import (
	"os"
	"time"
)

// AccessToken is a bearer token plus its expiry. A nil ExpiresAt means
// the token never expires (e.g. long-lived API keys) — the SDK fetches
// it once and never re-requests. Mirrors the core `AccessToken`.
type AccessToken struct {
	Token     string
	ExpiresAt *time.Time
}

// Credential abstracts a bearer credential for the REST API, mirroring
// the core `Credential` trait and the Python/Node `Credential` types.
// GetToken returns the current token and its expiry.
type Credential interface {
	GetToken() AccessToken
}

// ApiKeyCredential is a long-lived opaque API key — the only concrete
// Credential shipped today.
type ApiKeyCredential struct {
	key string
}

// NewApiKeyCredential creates an API-key credential. The key is sent as
// `Authorization: Bearer <key>`.
func NewApiKeyCredential(key string) *ApiKeyCredential {
	return &ApiKeyCredential{key: key}
}

// ApiKeyCredentialFromEnv reads the API key from BOXLITE_API_KEY. ok is
// false if the variable is unset or empty.
func ApiKeyCredentialFromEnv() (cred *ApiKeyCredential, ok bool) {
	key := os.Getenv("BOXLITE_API_KEY")
	if key == "" {
		return nil, false
	}
	return &ApiKeyCredential{key: key}, true
}

// GetToken returns the API key as a never-expiring bearer token.
func (c *ApiKeyCredential) GetToken() AccessToken {
	return AccessToken{Token: c.key, ExpiresAt: nil}
}

File: sdks/node/lib/credential.ts
/**
 * Credential abstraction for the BoxLite REST client.
 *
 * `Credential` is a structural interface: every credential class (today
 * `ApiKeyCredential`, a future `OAuthCredential`) structurally satisfies
 * it without a nominal `implements`. Callers type against `Credential`
 * and never change when new credential kinds are added.
 */

import type {
  ApiKeyCredential as NativeApiKeyCredential,
  JsAccessToken,
} from "./native-contracts.js";
import { getApiKeyCredential } from "./native.js";

/** A bearer token plus its expiry. */
export type AccessToken = JsAccessToken;

/** Anything that can produce a bearer token for `Boxlite.rest()`. */
export interface Credential {
  getToken(): AccessToken;
}

/**
 * Long-lived opaque API key credential. Concrete `Credential`.
 *
 * @example
 * ```ts
 * import { Boxlite, ApiKeyCredential } from '@boxlite/sdk';
 * const client = Boxlite.rest('http://localhost:8100',
 *                             new ApiKeyCredential('sk_...'));
 * ```
 */
export const ApiKeyCredential = getApiKeyCredential();

/** Instance type of {@link ApiKeyCredential}. */
export type ApiKeyCredential = NativeApiKeyCredential;
